src/plugins/vis_type_vega/public/expressions/helpers.ts

```diff
@@ -49,7 +50,22 @@ export const formatDatatable = (
   datatable.columns.forEach((column) => {
     // clean quotation marks from names in columns
     column.name = cleanString(column.name);
+    // clean ids to remove "." as that will cause vega to not process it correctly.
+    // This happens for different metric types
+    column.id = column.id.replaceAll('.', '-');
   });
+
+  // clean row keys to remove "." as that will cause vega to not process it correctly
+  const updatedRows: OpenSearchDashboardsDatatableRow[] = [];
+  datatable.rows.forEach((row) => {
+    const newRow = {};
+    for (const [key, value] of Object.entries(row)) {
+      const cleanKey = key.replaceAll('.', '-');
+      Object.assign(newRow, { [cleanKey]: value });
+    }
+    updatedRows.push(newRow);
+  });
+  datatable.rows = updatedRows;
```

Comment:
style nit suggestions - readability improvements are debatable, but you can avoid the temp variables with map and reduce.
```suggestion
  const updatedRows: OpenSearchDashboardsDatatableRow[] = datatable.rows.map((row) => (
    Object.entries(row).reduce((updatedRow, [key, value]) => {
      const cleanKey = key.replaceAll('.', '-');
      return Object.assign(updatedRow, { [cleanKey]: value });
    }, {});
  ));
  datatable.rows = updatedRows;
```